### ISSUE TITLE
Support partial updates when using `add` API

### ISSUE THREAD

**Issue by jakecraige** [realm/realm-cocoa#2644]
I'd love to see support for doing a partial update when using the `realm.add(..., update: true);` API. As it currently stands it seems this is supported by the `realm.create(..., update: true)` and not from `realm.add`. 

This is mentioned in the documentation like so when referencing `realm.create`:

> You can also partially update objects with primary keys by passing the subset of values you wish to update, along with the primary key:

I was able to track down test coverage of this feature(`.create`) to here
https://github.com/realm/realm-cocoa/blob/master/Realm/Tests/ObjectTests.m#L1568-L1577

I was able to write a failing spec describing the behavior and what I would like it to do here: https://github.com/realm/realm-cocoa/compare/master...jakecraige:jc-partial-update

**Comment by jpsim** [comment]
Hi Jake! Unfortunately, that's not really possible due to the way [`Realm.add()`](https://realm.io/docs/swift/latest/api/Classes/Realm.html#/s:FC10RealmSwift5Realm3addFS0_FTCS_6Object6updateSb_T_) works.

Since that function takes an Object, all its properties have values by definition (even if those values are sometimes `nil`). Swift (or Objective-C for that matter) doesn't provide any mechanism by which we can detect if a property was implicitly (at initialization) or explicitly set (later). We could introduce that tracking ourselves in the object, but that could add significant overhead in the general case, for questionable gain.

Besides, as far as I can tell, there are no scenarios in which you wouldn't be able to use [`Realm.create()`](https://realm.io/docs/swift/latest/api/Classes/Realm.html#/s:FC10RealmSwift5Realm6createuRdq_CS_6Object_FS0_FTMq_5valuePSs9AnyObject_6updateSb_q_) to accomplish the same thing.

**Comment by jakecraige** [comment]
What you said all makes sense. I've been toying with this locally to try and come up with a solution I'm happy with. I've thought of one solution I kind of like where we tell it exactly which keys to update. That way there doesn't need to be any tracking of that internally.

To work it in to my actual use case(with associations), it might look something like this:

```swift
let announcement = Announcement(value: [
  "id": 1
  "title": "New stuff",
  "user": PartialUpdate<User>(value: ["id": 2], updateKeys: ["id"])
])
realm.add(announcement, update: true)
```

The reason why I want to do all this is because I'm initializing objects from the JSON response from a server before persisting them so I'd prefer to have objects in memory that I can work with and persist them when I choose.

What do you think? Is something like this possible without core changes to realm?

**Comment by jakecraige** [comment]
A more in depth overview of the data-flow through my app looks like so(using Argo for the JSON parsing):

Models:

```swift
class User: Object {
    dynamic var id: Int = 0

    override static func primaryKey() -> String? {
        return "id"
    }
}

class Announcement: Object {
    dynamic var id: Int = 0
    dynamic var user: User?

    override static func primaryKey() -> String? {
        return "id"
    }
}

extension Announcement: Decodable {
    static func create(id id: Int, title: String, userId: Int) -> Announcement {
        return Announcement(value: [
            "id": id,
            // this is the part that breaks because
            // it wipes out existing user data when it saves
            "user": ["id": userId] 
        ])
    }

    static func decode(j: JSON) -> Decoded<Announcement> {
        return curry(Announcement.create)
            <^> j <|  "id"
            <*> j <|  "user_id"
    }
}

// this is major psuedo-code, in the real world I'm using thoughtbot's Swish 
// but it's not really related to this PR.
let announcementsJSON = requestData()
let announcements: Decoded<[Announcement]> = announcementsJSON.map { announcementJSON in
    return Announcement.decode(announcementJSON)
}

realm.write {
    realm.add(announcements, update: true)
}
```

Right now I don't need the database until the end which feels really nice. I could do my own find or initialize from the id for the user but I'd need the database for that, lol. 

Not sure if that's super helpful but it's actually what I'm trying to do in the real world so maybe you'll have other ideas about how I can accomplish this.

**Comment by jpsim** [comment]
Would this work?

```swift
let value = [
  "id": 1
  "title": "New stuff",
  "user": ["id": 2, "boolProperty": false]
]
realm.create(Announcement.self, value: value, update: true)
```

This will create or update both the `Announcement` object with an `id` of `1` and the `User` with an `id` of `2`.

**Comment by jpsim** [comment]
@jakecraige I'm just curious if my last comment was of any help to you. Ok to close this?

**Comment by jakecraige** [comment]
Yeah we can close it out. Your first comment about actually implementing a way to do this in core not being feasible addressed this.

I actually ended up removing Realm from the app I was working on because of the complexity of working around this to get what I wanted to happen. I'd love to add it back at some point but not being able to get around this without lots of extra boilerplate on my end was prohibitive so I opted to remove it and store stuff in memory for now.
